### PR TITLE
feat: V6 updates to toggle, library, setup, analytics and storage

### PR DESCRIPTION
**Pull request by stocaaro**
#### Description of changes:
Putting out an PR against the V6 branch updates.

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [x] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [x] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [x] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [x] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [x] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [x] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [x] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/fragments/lib-v1/storage/js/configureaccess.mdx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+```typescript
+Storage.configure({
+    customPrefix: {
+        guest: 'myPublicPrefix/',
+        protected: 'myProtectedPrefix/',
+        private: 'myPrivatePrefix/'
+    },
+    // ...
+})
+```
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/configureaccess.mdx:41]:
```suggestion
You can customize your key path by defining a prefix resolver:

```typescript
Amplify.configure(
    {},
    {
        Storage: {
            S3: {
                prefixResolver: async ({ accessLevel, targetIdentityId }) => {
                    if (accessLevel === 'guest') {
                        return 'myPublicPrefix/';
                    } else if (accessLevel === 'protected') {
                        return `myProtectedPrefix/${targetIdentityId}`;
                    } else {
                        return `myPrivatePrefix/${targetIdentityId}`;
                    }
                },
            },
        },
    }
);
```
```

File: src/fragments/lib-v1/storage/js/download.mdx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	url: URL;
+	expiresAt: Date;
+} = await getUrl({
+        key: filename,
+        options: { accessLevel: 'private' | 'protected' | 'guest' , // defaults to `guest` 
+        targetIdentityId?: string, // id of another user, if `level: protected`
+        validateObjectExistence: false,  // defaults to false
+        expiresIn: 20 // validity of the URL, in seconds. defaults to 900 (15 minutes) and maxes at 3600 (1 hour)
+        },
+    });
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/download.mdx:20]: This indentiaon seems to be off

File: src/fragments/lib-v1/storage/js/download.mdx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+You can check for the existence of a file in the storage category's `getUrl` API using the `validateObjectExistence` option. When this flag is enabled a `getUrl` call will return a pre-signed URL if the file exists and raise a `404` error if it does not. This allows you to check if an object exists during generating the presigned URL, which you can then use to download that object.
+
+```typescript
+// To check for existence of a file
+await getUrl({
+        key: filename,
+        options: {
+          validateObjectExistence: true,  // defaults to false
+        },
+    });
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/download.mdx:48]: Ditto indentation

File: src/fragments/lib-v1/storage/js/download.mdx [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+### getUrl expiry
+
+You can use `expiresIn` option to limit the availability of your URLs. This configuration returns the pre-signed URL that expires in 60 seconds:
+
+```javascript
+await Storage.getUrl('filename.txt', { expires: 60 });
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/download.mdx:56]:
```suggestion
await getUrl('filename.txt', { expiresIn: 60 });
```

File: src/fragments/lib-v1/storage/js/download.mdx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+const getUrlResult: {
+	url: URL;
+	expiresAt: Date;
+} = await getUrl({
+        key: filename,
+        options: { accessLevel: 'private' | 'protected' | 'guest' , // defaults to `guest` 
+        targetIdentityId?: string, // id of another user, if `level: protected`
+        validateObjectExistence: false,  // defaults to false
+        expiresIn: 20 // validity of the URL, in seconds. defaults to 900 (15 minutes) and maxes at 3600 (1 hour)
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/download.mdx:18]: Looks good for now, but I'd like to add a advanced use case section later showing how to get a long-live(7 days) presigned URL

File: src/fragments/lib-v1/storage/js/download.mdx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+			const downloadResult: { body: Pick<Body, 'blob' | 'json' | 'text'> } 
+			= await downloadData({ key: filename }).result;
+			if (downloadResult?.body?.text)
+				downloadResult?.body
+					?.text()
+					.then((result) => console.log(result));
+		} catch (error) {
+			console.log('Error : ', error)
+		}
+```
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/download.mdx:84]: Ditto indentation

File: src/fragments/lib-v1/storage/js/download.mdx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+To track the progress of your download, you can use `progressCallback`:
+
+```typescript
+// Download a file from s3 bucket
+  const { body, eTag } = await downloadData({ key, data: file, options: {
+  onProgress: (progress) {
+    console.log(`Downloaded: ${progress.loaded}/${progress.total}`);
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/download.mdx:92]:
```suggestion
    console.log(`Downloaded: ${progress.transferredBytes}/${progress.totalBytes}`);
```

File: src/fragments/lib-v1/storage/js/existing-resources.mdx [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+<Callout warning>
+  When you are not using Amplify CLI, adding existing Amazon S3 bucket to your application may require configuring bucket access permissions. e.g. Enabling read/write access to the cognito user pool that you are using with the Amplify Auth category.
+</Callout>
+
+```typescript
+
+Amplify.configure({
+  aws_user_files_s3_bucket_region: 'us-east-1', // (required) - Amazon S3 bucket region
+  aws_user_files_s3_bucket: 'testappa321fb103f1f2ae6a4d25d8cd2161728123152-dev' // (required) - Amazon S3 bucket URI
+});
```
Comment (AllanZhengYP) [inline src/fragments/lib-v1/storage/js/existing-resources.mdx:20]:
```suggestion
Amplify.configure({
    Storage: {
        S3: {
            region: 'us-east-1', // (required) - Amazon S3 bucket region
            bucket: 'testappa321fb103f1f2ae6a4d25d8cd2161728123152-dev' // (required) - Amazon S3 bucket URI
        }
    }
});
```

File: src/fragments/lib-v1/analytics/js/identifyuser.mdx [hunk 9]
```diff
@@ -0,0 +1,43 @@
+This call sends information about the current user (which could be unauthenticated or authenticated) to Amazon Pinpoint.
+
+You can provide `name`, `email` and `plan`, as well as location information with `AnalyticsUserProfile.Location`. You can also send additional custom attributes using  `AnalyticsProperties`.
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/identifyuser.mdx:3]:
```suggestion
This API sends information about the current user to Amazon Pinpoint.

Additional information such as the user's name, email, location, and device can be included by specifying the `UserProfile`. Custom attributes can also be included by setting `UserProfile.attributes`.

You can provide `name`, `email` and `plan`, as well as location information with `AnalyticsUserProfile.Location`. You can also send additional custom attributes using  `AnalyticsProperties`.
```

File: src/fragments/lib-v1/analytics/js/identifyuser.mdx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+If the user is signed in through [Amplify.Auth.signIn](/lib/auth/signin), then you can retrieve the current user's ID as shown below:
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/identifyuser.mdx:5]:
```suggestion
If the user was signed in through [signIn](/lib/auth/signin) you can retrieve the current user's ID as shown below:
```

File: src/fragments/lib-v1/analytics/js/identifyuser.mdx [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    email: "name@example.com",
+    plan: "plan",
+    location: location,
+    properties: properties
+}
+
+identifyUser({
+    userId: user.userId,
+    userProfile: userProfile
+});
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/identifyuser.mdx:38]:
```suggestion
async function sendUserData() {
    const user = await getCurrentUser(); 

    identifyUser({
        userId: user.userId,
        userProfile: userProfile
    });
}
```

File: src/fragments/lib-v1/analytics/js/identifyuser.mdx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+You can provide `name`, `email` and `plan`, as well as location information with `AnalyticsUserProfile.Location`. You can also send additional custom attributes using  `AnalyticsProperties`.
+
+If the user is signed in through [Amplify.Auth.signIn](/lib/auth/signin), then you can retrieve the current user's ID as shown below:
+
+```js
+import { identifyUser } from "aws-amplify/analytics";
+import { getCurrentUser } from "aws-amplify/auth";
+
+const user = await getCurrentUser(); 
+
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/identifyuser.mdx:12]: Await can't be used at module-scope

```suggestion
```

File: src/fragments/lib-v1/analytics/js/identifyuser.mdx [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+const location = {
+    latitude: 47.606209,
+    longitude: -122.332069,
+    postalCode: "98122",
+    city: "Seattle",
+    region: "WA",
+    country: "USA"
+};
+
+let properties = {
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/identifyuser.mdx:22]:
```suggestion
const attributes = {
        name: "username",
        email: "name@example.com",
        plan: "plan",
```

File: src/fragments/lib-v1/analytics/js/identifyuser.mdx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    "phoneNumber": "+11234567890",
+    "age": 25
+}
+
+let userProfile = {
+    name: "username",
+    email: "name@example.com",
+    plan: "plan",
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/identifyuser.mdx:30]:
```suggestion
```

File: src/fragments/lib-v1/analytics/js/record.mdx [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+```javascript
+import { record } from 'aws-amplify/analytics';
+
+record({ name: 'albumVisit' });
+```
+
+## Record a Custom Event with Attributes
+
+The `record` method lets you add additional attributes to an event. For example, to record _artist_ information with an _albumVisit_ event:
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/record.mdx:13]:
```suggestion
The `record` API lets you add additional attributes to an event. For example, to record _artist_ information with an _albumVisit_ event:
```

File: src/fragments/lib-v1/analytics/js/record.mdx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+```javascript
+import { record } from 'aws-amplify/analytics';
+
+record({
+  name: 'albumVisit',
+  // Attribute values must be strings
+  attributes: { genre: '', artist: '' }
+});
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/record.mdx:22]:
```suggestion
record({
    event: {
        name: 'albumVisit',
        attributes: { genre: '', artist: '' }
    }
});
```

File: src/fragments/lib-v1/analytics/js/record.mdx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+## Record Engagement Metrics
+
+Data can also be added to an event:
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/record.mdx:27]:
```suggestion
Metrics can also be added to an event:
```

File: src/fragments/lib-v1/analytics/js/record.mdx [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+```javascript
+import { record } from 'aws-amplify/analytics';
+
+record({
+  name: 'albumVisit',
+  attributes: {},
+  metrics: { minutesListened: 30 }
+});
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/record.mdx:36]:
```suggestion
record({
    record: {
        name: 'albumVisit',
        metrics: { minutesListened: 30 }
    }
});
```

File: src/fragments/lib-v1/analytics/js/record.mdx [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+Metric values must be a `Number` type such as a float or integer.
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/record.mdx:39]:
```suggestion
Metric values must be a `number` type such as a float or integer.
```

File: src/fragments/lib-v1/analytics/js/record.mdx [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+The Amazon Pinpoint & Kinesis providers send events in batches to optimize network bandwidth. However, events can be sent immediately using the `immediate` flag.
+
+```javascript
+import { record } from 'aws-amplify/analytics';
+
+record({
+  name: 'albumVisit',
+  immediate: true
+});
+```
```
Comment (jimblanc) [inline src/fragments/lib-v1/analytics/js/record.mdx:52]: This feature is not supported until GA:

```suggestion
```

File: src/fragments/lib-v1/start/getting-started/reactnative/setup.mdx [hunk 21]
```diff
@@ -0,0 +1,138 @@
+## Create a new project
```
Comment (jimblanc) [inline src/fragments/lib-v1/start/getting-started/reactnative/setup.mdx:1]: Should remove the React Native getting started guide as it's not supported for Dev Preview

File: src/fragments/lib-v1/start/getting-started/vanillajs/setup.mdx [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+    </style>
+  </head>
+
+  <body>
+    <div class="app">
+      <div class="app-header">
+        <h1>Welcome to Amplify</h1>
+      </div>
+      <div class="app-body">
+        <h1>Mutation Results</h1>
```
Comment (jimblanc) [inline src/fragments/lib-v1/start/getting-started/vanillajs/setup.mdx:101]: We'll need to re-work this as DataStore won't be available in preview. Okay doing it in a follow-up PR

File: src/pages/lib-v1/q/platform/[platform].mdx [hunk 23]
```diff
@@ -36,34 +36,34 @@ import flutter2 from '/src/fragments/lib-v1/flutter.mdx';
 
 ## Amplify JavaScript
 
+<Callout warning>
+
+Welcome to the developer preview of **Amplify JavaScript library V6**. This preview introduces functional interfaces, typescript improvements and improved NextJS SSR support  for Authentication, Analytics and Storage.
```
Comment (jimblanc) [inline src/pages/lib-v1/q/platform/[platform].mdx:41]:
```suggestion
Welcome to the developer preview of **Amplify JavaScript v6**. This preview introduces functional interfaces, typescript improvements and improved NextJS SSR support  for Authentication, Analytics, and Storage.
```

File: src/pages/lib-v1/q/platform/[platform].mdx [hunk 24]
```diff
@@ -36,34 +36,34 @@ import flutter2 from '/src/fragments/lib-v1/flutter.mdx';
...
+
+The team is working on the other categories for the V6 release. This developer preview is offered for you to try it out and provide feedback on the coming changes.
```
Comment (jimblanc) [inline src/pages/lib-v1/q/platform/[platform].mdx:43]:
```suggestion
The team is working on the other categories for the v6 release. This developer preview is offered for you to try it out and provide feedback on the coming changes. 
```

**Review by jimblanc — COMMENTED** [review]
Thank you so much for doing this! Dropped some initial feedback specifically on Analytics if we could merge in this PR, but this is a great base to start iterating on for more discrete changes.

**Review by jimblanc — APPROVED** [review]
Approving to unblock incremental changes & updates

File: src/pages/lib-v1/q/platform/[platform].mdx [hunk 25]
```diff
@@ -36,34 +36,34 @@ import flutter2 from '/src/fragments/lib-v1/flutter.mdx';
 
 ## Amplify JavaScript
 
+<Callout warning>
+
+Welcome to the developer preview of **Amplify JavaScript v6**. This preview introduces functional interfaces, typescript improvements and improved NextJS SSR support  for Authentication, Analytics, and Storage.
+
+The team is working on the other categories for the v6 release. This developer preview is offered for you to try it out and provide feedback on the coming changes. 
+
+V6_TODO: Retool this callout and link to RFC
```
Comment (ashwinkumar6) [inline src/pages/lib-v1/q/platform/[platform].mdx:45]: I'm wondering if we should we comment this line

Comment (stocaaro) [reply src/pages/lib-v1/q/platform/[platform].mdx:45]: Leaving it in so that it gets further attention before we merge it.

**Pull request merged**